### PR TITLE
Fixed #34409 -- Doc'd limitation of dictfetchall() and namedtuplefetchall() examples.

### PR DESCRIPTION
**Pull request by jesperolsson-se**

### CODE REVIEW

File: docs/topics/db/sql.txt [hunk 1]
```diff
@@ -323,7 +323,12 @@ small performance and memory cost, you can return results as a ``dict`` by
 using something like this::
 
     def dictfetchall(cursor):
-        "Return all rows from a cursor as a dict"
+        """Return all rows from a cursor as a dict
+
+        Column names must be unique or the dict value will be overwritten.
+        Since the cursor cannot distinguish columns from different tables,
+        JOINs are not exempt.
```
Comment (felixxm) [inline docs/topics/db/sql.txt:330]: I would remove ``JOINs are not exempt.`` and move it outside of the docstring. For example:
```diff
diff --git a/docs/topics/db/sql.txt b/docs/topics/db/sql.txt
index 94a724af04..57405b1d0a 100644
--- a/docs/topics/db/sql.txt
+++ b/docs/topics/db/sql.txt
@@ -341,6 +341,9 @@ immutable and accessible by field names or indices, which might be useful::
         nt_result = namedtuple("Result", [col[0] for col in desc])
         return [nt_result(*row) for row in cursor.fetchall()]
 
+The ``dictfetchall()`` and ``namedtuplefetchall()`` examples assume unique
+column names, since a cursor cannot distinguish columns from different tables.
+
 Here is an example of the difference between the three:
 
 .. code-block:: pycon

```

**Review by felixxm — COMMENTED** [review]
@jesperolsson-se Thanks :+1:

File: docs/topics/db/sql.txt (hunk 1 above)
Comment (jesperolsson-se) [reply docs/topics/db/sql.txt:330]: There are a couple of benefits to using the docstring that I'd prefer not to miss out on. Essentially, I view this information as relating directly to the code and should be kept as close to it as possible.

Take [this StackOverflow answer](https://stackoverflow.com/a/14294314) as an example. The code snippet is copied verbatim but the surrounding documentation isn't. Reading that answer, finding a copy in one's code base or skimming the documentation does not reveal the hidden assumption. In fact, the current docstring explicitly says "all rows". To catch the limitation, you'd have to be intimately familiar with the workings of `cursor`.

Similarly, once the documentation starts describing `namedtuplefetchall`, it isn't obvious that we'll circle back and talk more about `dictfetchall`. The reader who is solely interested in this method would need to realize that there's more information later on.

Put simply, docstrings are excellent for documenting at the code level and I believe that we'll save our users some trouble, by placing this caveat there.

Comment (felixxm) [reply docs/topics/db/sql.txt:330]: > Take [this StackOverflow answer](https://stackoverflow.com/a/14294314) as an example. The code snippet is copied verbatim but the surrounding documentation isn't.

We cannot take responsibility for how and where folks copy Django docs.

> To catch the limitation, you'd have to be intimately familiar with the workings of `cursor`.

No, You'd have to read the docs this snippet comes from.

> Put simply, docstrings are excellent for documenting at the code level and I believe that we'll save our users some trouble, by placing this caveat there.

Following your recommendation we'd have to copy all warnings, admonitions, etc. to the comments in examples :upside_down_face:

Comment (felixxm) [reply docs/topics/db/sql.txt:330]: We can also have both, e.g.
```diff
diff --git a/docs/topics/db/sql.txt b/docs/topics/db/sql.txt
index 94a724af04..217ecb987d 100644
--- a/docs/topics/db/sql.txt
+++ b/docs/topics/db/sql.txt
@@ -323,7 +323,10 @@ small performance and memory cost, you can return results as a ``dict`` by
 using something like this::
 
     def dictfetchall(cursor):
-        "Return all rows from a cursor as a dict"
+        """
+        Return all rows from a cursor as a dict.
+        Assume the column names are unique.
+        """
         columns = [col[0] for col in cursor.description]
         return [dict(zip(columns, row)) for row in cursor.fetchall()]
 
@@ -336,11 +339,17 @@ immutable and accessible by field names or indices, which might be useful::
 
 
     def namedtuplefetchall(cursor):
-        "Return all rows from a cursor as a namedtuple"
+        """
+        Return all rows from a cursor as a namedtuple.
+        Assume the column names are unique.
+        """
         desc = cursor.description
         nt_result = namedtuple("Result", [col[0] for col in desc])
         return [nt_result(*row) for row in cursor.fetchall()]
 
+The ``dictfetchall()`` and ``namedtuplefetchall()`` examples assume unique
+column names, since a cursor cannot distinguish columns from different tables.
+
 Here is an example of the difference between the three:
 
 .. code-block:: pycon

```

**Comment by felixxm** [comment]
@jesperolsson-se Welcome aboard :boat:
